In [ ]:
# import libraries 
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 

In [ ]:
# Data data preprocessing step
users_behavior = pd.read_csv('/home/susan/machine_learning_project/Data/users_behavior.csv')
print(users_behavior.info())
print()
print('====='*30)
print(users_behavior.head())
print()
print(users_behavior.sample(random_state= 1234))
print('====='*30)
# Normalize column names
users_behavior.columns = users_behavior.columns.str.strip().str.lower().str.replace(' ', '_')
# missing data
print(users_behavior.isna().sum())
print()
print('====='*30)
# duplicated data
print(users_behavior.duplicated().sum())
print()
print('====='*30)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

      calls  minutes  messages   mb_used  is_ultra
1758   77.0   496.76       0.0  17765.34         0
calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

0



In [18]:
# Step 2 Ml processing 
#Libraries
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.model_selection import StratifiedKFold, cross_val_score

In [13]:
# Create Feature and targets
features = users_behavior.drop(['is_ultra'], axis= 1)
target = users_behavior['is_ultra']
print(features.shape)
print(target.shape)
# Step 1: Split into "temp" (80%) and "test" (20%)
X_train, X_test, y_train ,y_test = train_test_split(features, target, test_size= 0.2, random_state = 42)

(3214, 4)
(3214,)


In [15]:
# DecisionTree Classifier

model = DecisionTreeClassifier(random_state= 42)

model.fit(X_train , y_train)
predictions = model.predict(X_test)
accuracy = accuracy_score(y_test, predictions)
precision = precision_score(y_test, predictions)
recall = recall_score(y_test, predictions)
f1 = f1_score(y_test, predictions)
confusion_mat = confusion_matrix(y_test, predictions)

print("=== DECISION TREE PERFORMANCE ===")
print(f"Accuracy:  {accuracy:.1%}")
print(f"Precision: {precision:.1%}")
print(f"Recall:    {recall:.1%}")
print(f"F1 Score:  {f1:.1%}")

print('======'*30)
#Create a tuned Decision Tree with optimized hyperparameters
model_tuned = DecisionTreeClassifier(max_depth=6, 
                                    min_samples_split=10,
                                    min_samples_leaf=5,
                                    max_features='sqrt',
                                    random_state=42)
# Train the model
model_tuned.fit(X_train, y_train)

# Make predictions
predictions_tuned = model_tuned.predict(X_test)

# Calculate performance metrics
accuracy_tuned = accuracy_score(y_test, predictions_tuned)
precision_tuned = precision_score(y_test, predictions_tuned)
recall_tuned = recall_score(y_test, predictions_tuned)
f1_tuned = f1_score(y_test, predictions_tuned)


print("=== TUNED DECISION TREE PERFORMANCE ===")
print(f"Accuracy:  {accuracy_tuned:.1%}")
print(f"Precision: {precision_tuned:.1%}")
print(f"Recall:    {recall_tuned:.1%}")
print(f"F1 Score:  {f1_tuned:.1%}")
print()
print("Hyperparameters used:")
print(f"Max Depth: {model_tuned.max_depth}")
print(f"Min Samples Split: {model_tuned.min_samples_split}")
print(f"Min Samples Leaf: {model_tuned.min_samples_leaf}")
print(f"Max Features: {model_tuned.max_features}")






=== DECISION TREE PERFORMANCE ===
Accuracy:  74.7%
Precision: 56.9%
Recall:    54.8%
F1 Score:  55.8%
=== TUNED DECISION TREE PERFORMANCE ===
Accuracy:  79.9%
Precision: 79.2%
Recall:    42.6%
F1 Score:  55.4%

Hyperparameters used:
Max Depth: 6
Min Samples Split: 10
Min Samples Leaf: 5
Max Features: sqrt


In [17]:
# RandomForestClassifier

model_random = RandomForestClassifier(random_state= 42)

model_random.fit(X_train , y_train)
predictions_random =model_random.predict(X_test)
accuracy = accuracy_score(y_test, predictions_random)
precision = precision_score(y_test, predictions_random)
recall = recall_score(y_test, predictions_random)
f1 = f1_score(y_test, predictions_random)
confusion_mat = confusion_matrix(y_test, predictions_random)

print("=== RandomForestClassifier PERFORMANCE  ===")
print(f"Accuracy:  {accuracy:.1%}")
print(f"Precision: {precision:.1%}")
print(f"Recall:    {recall:.1%}")
print(f"F1 Score:  {f1:.1%}")

print('======'*30)
#Create a tuned RandomForestClassifier with optimized hyperparameters
model_optimize = RandomForestClassifier( n_estimators = 200, # number of trees ( more better, but slower )
                                       max_depth = 10, # Indivitual tree depth
                                       min_samples_split = 5,#Tree split criteria
                                      min_samples_leaf=2, # 
                                       max_features= 'sqrt', # features per tree(default for classification)
                                       bootstrap= True, # sampler with replacement
                                       random_state = 42)

model_optimize.fit(X_train , y_train)
predictions_optimized = model_optimize.predict(X_test)
accuracy_optimized = accuracy_score(y_test, predictions_optimized)
precision_optimized = precision_score(y_test, predictions_optimized)
recall_optimized = recall_score(y_test, predictions_optimized)
f1_optimized = f1_score(y_test, predictions_optimized)
confusion_mat_optimized = confusion_matrix(y_test, predictions_optimized)

print("=== OPTIMIZED RANDOM FOREST PERFORMANCE ===")
print(f"Accuracy:  {accuracy_optimized:.1%}")
print(f"Precision: {precision_optimized:.1%}")
print(f"Recall:    {recall_optimized:.1%}")
print(f"F1 Score:  {f1_optimized:.1%}")
print()
print("Hyperparameters used:")
print(f"Number of trees: {model_optimize.n_estimators}")
print(f"Max Depth: {model_optimize.max_depth}")
print(f"Min Samples Split: {model_optimize.min_samples_split}")
print(f"Min Samples Leaf: {model_optimize.min_samples_leaf}")
print(f"Max Features: {model_optimize.max_features}")
print()
print('======'*30)


=== RandomForestClassifier PERFORMANCE  ===
Accuracy:  81.2%
Precision: 73.4%
Recall:    55.9%
F1 Score:  63.4%
=== OPTIMIZED RANDOM FOREST PERFORMANCE ===
Accuracy:  81.3%
Precision: 77.4%
Recall:    51.1%
F1 Score:  61.5%

Hyperparameters used:
Number of trees: 200
Max Depth: 10
Min Samples Split: 5
Min Samples Leaf: 2
Max Features: sqrt



In [ ]:
# This code compares two Random Forest models:
# Baseline model (default settings)
#Tuned model (optimized hyperparameters)
# The purpose is to determine whether hyperparameter 
# tuning improves model performance.

# Step 1. Create the Baseline Model
rf_base = RandomForestClassifier(random_state=42)

# Step 2. Create the Tuned Model
rf_tuned = RandomForestClassifier(
    n_estimators=200,# 200 decision tree
    max_depth=10,  # max deepth of each tree
    min_samples_split=5, # for splitting need at least 5 samples
    min_samples_leaf=2, # each leaf must contain at least 2 samples
    max_features='sqrt',
    bootstrap=True,
    random_state=42
)
# Step 3. Create Cross Validation
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
 # What is this?

# Instead of one train/test split,the dataset is divided into 5 folds.
# Each fold becomes the validation set once.
# Training happens 5 times.


# Step 4. Cross Validation Scores
base_cv_f1 = cross_val_score(rf_base, features, target, cv=cv, scoring='f1').mean()
tuned_cv_f1 = cross_val_score(rf_tuned, features, target, cv=cv, scoring='f1').mean()

# Step 5. Train Final Models
rf_base.fit(X_train, y_train)
rf_tuned.fit(X_train, y_train)

#Step 5. Train Final Models
base_train_f1 = f1_score(y_train, rf_base.predict(X_train))
base_test_f1 = f1_score(y_test, rf_base.predict(X_test))

tuned_train_f1 = f1_score(y_train, rf_tuned.predict(X_train))
tuned_test_f1 = f1_score(y_test, rf_tuned.predict(X_test))

print("=== Cross-Validation F1 (5-fold) ===")
print(f"Baseline RF CV F1: {base_cv_f1:.4f}")
print(f"Tuned RF CV F1:    {tuned_cv_f1:.4f}")

print("\n=== Train vs Test F1 ===")
print(f"Baseline RF -> Train: {base_train_f1:.4f}, Test: {base_test_f1:.4f}")
print(f"Tuned RF    -> Train: {tuned_train_f1:.4f}, Test: {tuned_test_f1:.4f}")

=== Cross-Validation F1 (5-fold) ===
Baseline RF CV F1: 0.6245
Tuned RF CV F1:    0.6245

=== Train vs Test F1 ===
Baseline RF -> Train: 1.0000, Test: 0.6344
Tuned RF    -> Train: 0.7544, Test: 0.6154


**Both models achieved the same average F1 score (0.6245) during 5-fold cross-validation.This suggests that, on average, hyperparameter tuning did not improve the model's expected performance.**

**"The baseline Random Forest achieved a perfect F1 score on the training data but a much lower score on the test data. This large gap indicates that the model overfit the training data and did not generalize well."**

**"The tuned model produced a lower training F1 score, indicating reduced model complexity. The smaller gap between the training and test scores suggests that tuning reduced overfitting and improved generalization, although the overall test performance did not improve."**

| Model    | Train F1 | Test F1 | Interpretation                                             |
| -------- | -------: | ------: | ---------------------------------------------------------- |
| Baseline |   1.0000 |  0.6344 | Strong overfitting                                         |
| Tuned    |   0.7544 |  0.6154 | Less overfitting but slightly lower predictive performance |



=========================================================================

**"I compared a baseline Random Forest model with a tuned version using optimized hyperparameters. The optimized model slightly improved accuracy (81.3% vs. 81.2%) and precision (77.4% vs. 73.4%), meaning its positive predictions were more reliable. However, recall decreased from 55.9% to 51.1%, indicating that it missed more positive cases. Because the F1 score also decreased from 63.4% to 61.5%, the baseline model provided the best overall balance between precision and recall. Therefore, I selected the baseline Random Forest as the final model."**

========================================================================